File: AI-Fitness/frontend/movement_copilot.js
/**
 * FitQuest Human Movement Intelligence Engine — Phase 5: MOVEMENT COPILOT™
 * Real-Time Biomechanical Degradation Detection, Context-Aware Coaching, & Closed-Loop Recovery HUD.
 */

// ---------------------------------------------------------------------------
// Global Copilot Client State
// ---------------------------------------------------------------------------
let copilotSessionId = null;
let copilotExerciseId = "1";
let copilotExerciseName = "Exercise";
let copilotDnaLimiter = null;
let copilotAdaptiveTempo = null;
let isCopilotActive = false;
let lastCopilotFrameTimestamp = 0;
const COPILOT_THROTTLE_MS = 300; // Send telemetry every 300ms for stable low-latency coaching

let copilotEventTimeline = [];

/**
 * Initializes Copilot engine listeners and hooks
 */
function initializeCopilot() {
  console.log('[FitQuest Copilot]: Initialized Movement Copilot™ Engine.');
}

/**
 * Starts a new active Copilot session
 */
function startCopilot(sessionId, exercise, options = {}) {
  copilotSessionId = sessionId || `copilot_${Date.now()}`;
  copilotExerciseId = String(exercise?.id || "1");
  copilotExerciseName = exercise?.name || "Exercise";
  copilotDnaLimiter = options.dnaLimiter || null;
  copilotAdaptiveTempo = options.adaptiveTempo || null;
  isCopilotActive = true;
  lastCopilotFrameTimestamp = 0;
  copilotEventTimeline = [];

  const copilotCard = document.getElementById('movementCopilotCard');
  if (copilotCard) {
    copilotCard.style.display = 'flex';
  }

  renderCopilotHUD({
    status: 'IDLE',
    exercise: copilotExerciseName,
    dimension_label: 'Overall Movement',
    score: 80.0,
    message: 'Calibrating camera & pose sensors...',
    coaching_cue: 'Position yourself in full view to begin.',
    movement_phase: 'PREPARE',
    current_metrics: {
      range_of_motion: 80.0,
      movement_stability: 80.0,
      tempo_control: 80.0,
      repetition_consistency: 80.0
    }
  });
}

/**
 * Feeds live frame telemetry into Copilot Engine (throttled)
 */
async function processCopilotTelemetry(telemetry) {
  if (!isCopilotActive || !copilotSessionId || !telemetry) return;

  const now = Date.now();
  if ((now - lastCopilotFrameTimestamp) < COPILOT_THROTTLE_MS) {
    return;
  }
  lastCopilotFrameTimestamp = now;

  const token = typeof getAuthToken === 'function' ? getAuthToken() : (localStorage.getItem('fitquest_token') || localStorage.getItem('token'));
  if (!token) return;

  try {
    const payload = {
      session_id: copilotSessionId,
      exercise_id: copilotExerciseId,
      rep_count: telemetry.rep_count || 0,
      form_score: telemetry.form_score || 0.0,
      state: telemetry.state || 'START',
      primary_angle: telemetry.primary_angle || null,
      left_angle: telemetry.left_angle || null,
      right_angle: telemetry.right_angle || null,
      torso_angle: telemetry.torso_angle || null,
      valid: telemetry.valid !== false,
      feedback_code: telemetry.feedback_code || 'GOOD_FORM',
      dna_primary_limiter: copilotDnaLimiter,
      adaptive_tempo_target: copilotAdaptiveTempo
    };

    const res = await fetch(`${API_BASE}/movement-intelligence/copilot/frame`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${token}`
      },
      body: JSON.stringify(payload)
    });

    if (res.ok) {
      const copilotData = await res.json();
      renderCopilotHUD(copilotData);
    }
  } catch (err) {
    console.warn('[FitQuest Copilot]: Telemetry ingestion warning:', err);
  }
}

/**
 * Renders the real-time Copilot HUD Card
 */
function renderCopilotHUD(data) {
  if (!data) return;

  // 1. Status Pill & Badge
  const statusBadge = document.getElementById('copilotStatusBadge');
  const statusText = document.getElementById('copilotStatusText');
  if (statusBadge && statusText) {
    statusBadge.className = 'copilot-status-badge';
    if (data.status === 'RECOVERED') {
      statusBadge.classList.add('status-recovered');
      statusText.innerHTML = '<i class="fa-solid fa-check-circle"></i> RECOVERED';
    } else if (data.status === 'WARNING' || data.severity === 'WARNING') {
      statusBadge.classList.add('status-warning');
      statusText.innerHTML = '<i class="fa-solid fa-triangle-exclamation"></i> DEGRADATION';
    } else if (data.status === 'CRITICAL' || data.severity === 'CRITICAL') {
      statusBadge.classList.add('status-critical');
      statusText.innerHTML = '<i class="fa-solid fa-circle-exclamation"></i> CRITICAL FORM';
    } else {
      statusBadge.classList.add('status-monitoring');
      statusText.innerHTML = '<span class="copilot-pulse-dot"></span> MONITORING';
    }
  }

  // 2. Primary Dimension Name & Score
  const dimNameEl = document.getElementById('copilotDimName');
  const dimScoreEl = document.getElementById('copilotDimScore');
  const dimBarEl = document.getElementById('copilotDimBar');
  const deltaBadgeEl = document.getElementById('copilotDeltaBadge');

  if (dimNameEl) dimNameEl.innerText = data.dimension_label || 'Movement Stability';
  if (dimScoreEl) dimScoreEl.innerText = `${Math.round(data.score || 80)} / 100`;
  if (dimBarEl) dimBarEl.style.width = `${Math.min(100, Math.max(10, data.score || 80))}%`;

  if (deltaBadgeEl) {
    if (data.delta && Math.abs(data.delta) >= 1.0) {
      deltaBadgeEl.style.display = 'inline-flex';
      const sign = data.delta > 0 ? '+' : '';
      deltaBadgeEl.innerText = `${sign}${data.delta.toFixed(1)} pts`;
      deltaBadgeEl.className = data.delta > 0 ? 'copilot-delta-pill delta-pos' : 'copilot-delta-pill delta-neg';
    } else {
      deltaBadgeEl.style.display = 'none';
    }
  }

  // 3. AI Coach Cue & Phase Pill
  const cueEl = document.getElementById('copilotCueText');
  const phasePill = document.getElementById('copilotPhasePill');

  if (cueEl) cueEl.innerText = data.coaching_cue || 'Maintain steady cadence and control.';
  if (phasePill) {
    phasePill.innerText = data.movement_phase || 'STEADY';
  }

  // 4. Mini Metric Progress Bars
  const metrics = data.current_metrics || {};
  updateMiniMetric('copilotMiniStab', metrics.movement_stability || 80);
  updateMiniMetric('copilotMiniTempo', metrics.tempo_control || 80);
  updateMiniMetric('copilotMiniRom', metrics.range_of_motion || 80);
  updateMiniMetric('copilotMiniCons', metrics.repetition_consistency || 80);

  // 5. Timeline Events
  if (data.recent_events && data.recent_events.length > 0) {
    renderCopilotTimeline(data.recent_events);
  }
}

function updateMiniMetric(elementId, value) {
  const bar = document.getElementById(`${elementId}Bar`);
  const val = document.getElementById(`${elementId}Val`);
  const v = Math.round(value || 80);
  if (bar) bar.style.width = `${Math.min(100, Math.max(5, v))}%`;
  if (val) val.innerText = `${v}`;
}

function renderCopilotTimeline(events) {
  const container = document.getElementById('copilotTimelineStream');
  if (!container || !events) return;

  container.innerHTML = events.slice(-3).map(ev => {
    let icon = '<i class="fa-solid fa-circle-info" style="color: var(--accent-cyan);"></i>';
    if (ev.type && ev.type.includes('RECOVERY')) {
      icon = '<i class="fa-solid fa-circle-check" style="color: var(--status-success);"></i>';
    } else if (ev.type && ev.type.includes('CRITICAL')) {
      icon = '<i class="fa-solid fa-circle-exclamation" style="color: #ef4444;"></i>';
    } else if (ev.type && ev.type.includes('INTERVENTION')) {
      icon = '<i class="fa-solid fa-triangle-exclamation" style="color: var(--status-warning);"></i>';
    }
    return `
      <div class="copilot-timeline-item">
        <span class="copilot-tl-rep">REP ${String(ev.rep || 0).padStart(2, '0')}</span>
        <span class="copilot-tl-icon">${icon}</span>
        <span class="copilot-tl-msg">${escapeHTML(ev.message || '')}</span>
      </div>
    `;
  }).join('');
}

/**
 * Resets Copilot on exercise transition / new set
 */
async function resetCopilot(newExercise) {
  if (newExercise) {
    copilotExerciseId = String(newExercise.id || "1");
    copilotExerciseName = newExercise.name || "Exercise";
  }
  const token = typeof getAuthToken === 'function' ? getAuthToken() : (localStorage.getItem('fitquest_token') || localStorage.getItem('token'));
  if (token && copilotSessionId) {
    try {
      await fetch(`${API_BASE}/movement-intelligence/copilot/reset`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${token}`
        },
        body: JSON.stringify({
          session_id: copilotSessionId,
          exercise_id: copilotExerciseId
        })
      });
    } catch (e) {
      console.warn('[FitQuest Copilot]: Reset error:', e);
    }
  }
}

/**
 * Stops Copilot session on workout completion
 */
function stopCopilot() {
  isCopilotActive = false;
  const copilotCard = document.getElementById('movementCopilotCard');
  if (copilotCard) {
    copilotCard.style.display = 'none';
  }
}

/**
 * Fetches and renders Post-Workout Copilot Session Summary
 */
async function loadPostWorkoutCopilotSummary(sessionId) {
  const targetSessionId = sessionId || copilotSessionId;
  const summaryCard = document.getElementById('resMovementCopilotCard');
  if (!summaryCard || !targetSessionId) return;

  const token = typeof getAuthToken === 'function' ? getAuthToken() : (localStorage.getItem('fitquest_token') || localStorage.getItem('token'));
  if (!token) return;

  try {
    const res = await fetch(`${API_BASE}/movement-intelligence/copilot/summary/${targetSessionId}`, {
      headers: {
        'Authorization': `Bearer ${token}`,
        'Accept': 'application/json'
      }
    });

    if (res.ok) {
      const summary = await res.json();
      renderCopilotSummaryCard(summary);
    }
  } catch (err) {
    console.warn('[FitQuest Copilot]: Failed to load post-workout summary:', err);
  }
}

/**
 * Renders Post-Workout Intelligence Summary Card
 */
function renderCopilotSummaryCard(summary) {
  const summaryCard = document.getElementById('resMovementCopilotCard');
  if (!summaryCard || !summary) return;

  summaryCard.style.display = 'flex';

  const intEl = document.getElementById('resCopilotInterventions');
  const corEl = document.getElementById('resCopilotCorrections');
  const rateEl = document.getElementById('resCopilotRecoveryRate');
  const bigEl = document.getElementById('resCopilotBiggestRecovery');
  const freqEl = document.getElementById('resCopilotMostFreqLimiter');
  const verEl = document.getElementById('resCopilotVerdict');

  if (intEl) intEl.innerText = summary.total_interventions ?? 0;
  if (corEl) corEl.innerText = summary.successful_corrections ?? 0;
  if (rateEl) rateEl.innerText = `${(summary.recovery_rate_pct ?? 100).toFixed(0)}%`;

  if (bigEl) {
    if (summary.biggest_recovery) {
      bigEl.innerText = `${summary.biggest_recovery.label} (+${summary.biggest_recovery.gain_pts} pts)`;
    } else {
      bigEl.innerText = 'Optimal Consistency';
    }
  }

  if (freqEl) {
    if (summary.most_frequent_limiter) {
      freqEl.innerText = `${summary.most_frequent_limiter.label} (${summary.most_frequent_limiter.count}x)`;
    } else {
      freqEl.innerText = 'None (Stable)';
    }
  }

  if (verEl) {
    verEl.innerText = summary.verdict || 'Flawless execution throughout your workout set.';
  }
}

// Auto-initialize on load
document.addEventListener('DOMContentLoaded', () => {
  initializeCopilot();
});

File: AI-Fitness/backend/api/movement_copilot.py
"""
FitQuest Human Movement Intelligence Engine — Phase 5: MOVEMENT COPILOT™ API
Lightweight real-time endpoint for session telemetry ingestion, degradation alerts, and recovery confirmation.
"""

from typing import Optional, Dict, Any, Union
from pydantic import BaseModel, Field
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from backend.database import get_db
from backend.api.auth import get_current_user_id
from backend.services.movement_copilot import movement_copilot_engine


router = APIRouter(prefix="/movement-intelligence/copilot", tags=["Movement Copilot™"])


class CopilotFrameRequest(BaseModel):
    session_id: str = Field(..., description="Active workout session ID")
    exercise_id: Union[int, str] = Field(..., description="Exercise ID (1-20)")
    rep_count: int = Field(default=0, ge=0, description="Current repetition count")
    form_score: float = Field(default=0.0, ge=0.0, le=100.0, description="Current form score (0-100)")
    state: Optional[str] = Field(default="START", description="Exercise tracker kinematic state")
    primary_angle: Optional[float] = Field(default=None, description="Primary joint angle (degrees)")
    left_angle: Optional[float] = Field(default=None, description="Left joint angle (degrees)")
    right_angle: Optional[float] = Field(default=None, description="Right joint angle (degrees)")
    torso_angle: Optional[float] = Field(default=None, description="Torso inclination angle (degrees)")
    valid: bool = Field(default=True, description="Landmarks validity flag")
    feedback_code: Optional[str] = Field(default="GOOD_FORM", description="Tracker feedback code")
    dna_primary_limiter: Optional[str] = Field(default=None, description="Phase 4 Movement DNA primary limiter")
    adaptive_tempo_target: Optional[str] = Field(default=None, description="Phase 3 Adaptive Training tempo target (e.g. '3-1-2')")


class CopilotResetRequest(BaseModel):
    session_id: str = Field(..., description="Workout session ID to reset")
    exercise_id: Optional[Union[int, str]] = Field(default=None, description="New exercise ID if switching")


@router.post("/frame", summary="Process Live Movement Telemetry with Movement Copilot™")
def process_copilot_frame(
    payload: CopilotFrameRequest,
    current_user_id: int = Depends(get_current_user_id)
):
    """
    Real-time endpoint evaluating rolling movement telemetry.
    Detects biomechanical degradation, context-aware coaching cues, and closed-loop recovery.
    """
    try:
        result = movement_copilot_engine.process_frame_telemetry(
            session_id=payload.session_id,
            user_id=current_user_id,
            exercise_id=str(payload.exercise_id),
            rep_count=payload.rep_count,
            form_score=payload.form_score,
            state=payload.state or "START",
            primary_angle=payload.primary_angle,
            left_angle=payload.left_angle,
            right_angle=payload.right_angle,
            torso_angle=payload.torso_angle,
            valid=payload.valid,
            feedback_code=payload.feedback_code or "GOOD_FORM",
            dna_primary_limiter=payload.dna_primary_limiter,
            adaptive_tempo_target=payload.adaptive_tempo_target
        )
        return result
    except PermissionError as pe:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=str(pe))
    except Exception as e:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=f"Copilot processing error: {e}")


@router.post("/reset", summary="Reset Copilot State for Session / Exercise Transition")
def reset_copilot_session(
    payload: CopilotResetRequest,
    current_user_id: int = Depends(get_current_user_id)
):
    """
    Resets Copilot rolling window and active state on exercise change.
    """
    try:
        success = movement_copilot_engine.reset_session(
            session_id=payload.session_id,
            user_id=current_user_id,
            new_exercise_id=str(payload.exercise_id) if payload.exercise_id else None
        )
        return {"status": "success", "reset": success, "session_id": payload.session_id}
    except PermissionError as pe:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=str(pe))
    except Exception as e:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


@router.get("/summary/{session_id}", summary="Get Post-Workout Movement Copilot Session Intelligence")
def get_copilot_summary(
    session_id: str,
    current_user_id: int = Depends(get_current_user_id)
):
    """
    Fetches the final Copilot session intelligence summary for the completed workout.
    """
    try:
        summary = movement_copilot_engine.get_session_summary(
            session_id=session_id,
            user_id=current_user_id
        )
        return summary
    except Exception as e:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))
